### DISCUSSION TITLE
Pimcore Copilot: Implement Custom Environment Variable Types

### DISCUSSION THREAD

**Discussion by Chauhan-Mukesh** [pimcore/pimcore#17100] [General]
Hi Pimcore community,

I'm working on a Pimcore project (version 11.2.3.2) and I'm trying to implement custom environment variable types in Pimcore Copilot for a Select field.  My goal is to allow users to define `optionsProviderClass` and `optionsProviderData` so I can dynamically generate the select options. This dynamic select will then be used in a Copilot Automation Action to generate a variant object.

Has anyone successfully done this before? If so, could you please share some guidance or code examples on how to:

1.  **Create the custom environment variable type for the Select field.**
2.  **Implement the `optionsProviderClass` and `optionsProviderData` logic.**
3.  **Use this dynamic Select field in a Copilot Automation Action to create a variant object.**

Any help or pointers in the right direction would be greatly appreciated!

This much of configuration I have done.
```
<?php

namespace App\Copilot\EnvironmentVariables\Types;

use Pimcore\Bundle\CopilotBundle\AutomationAction\Configuration\Environment\AbstractEnvironment;

final class DynamicSelect extends AbstractEnvironment
{
    public function getType(): string
    {
        return 'select';
    }

    public function getLabel(): string
    {
        return 'Dynamic Select';
    }

    public function getIcon(): string
    {
        return 'pimcore_icon_select';
    }

    protected function configureEnvironment(): void
    {
        $this->environmentConfiguration->setRequired('required');
        $this->environmentConfiguration->setDefault('required', false);
        $this->environmentConfiguration->setAllowedTypes('required', 'bool');

        $this->environmentConfiguration->setRequired('optionProviderClass');
        $this->environmentConfiguration->setDefault('required', false);
        $this->environmentConfiguration->setAllowedTypes('optionProviderClass', 'string');

        $this->environmentConfiguration->setRequired('optionProviderData');
        $this->environmentConfiguration->setDefault('required', false);
        $this->environmentConfiguration->setAllowedTypes('optionProviderData', 'string');
    }
}
```

```
services:
    App\Copilot\EnvironmentVariables\Types\DynamicSelect:
        tags:
            - { name: 'pimcore.copilot.automation_action.configuration_environment_type' }
```
Thanks in advance!

**Comment by fashxp** [comment]
Hi,
your custom type probably needs to overwrite `public function resolveConfiguration(): array` and inject the dynamic options provider data there into the `values` key of the returned array - as this is then used in the frontend to show the options. 

Cheers
